File: Week3/Code/get_TreeHeight.R
##################################################################
## Tree heights script taking input files from the command line ##
##################################################################

#Take a csv file from the command line
#Output to results file but input file name in the output file name
#NOTE: strip .csv
#use commandArgs

#clear environment
rm(list = ls())
#trees <- read.csv("../Data/trees.csv", header = TRUE)
#This function calculates heights of tree given the distance of each tree,
#form its base and angle to its top
#uses the trigonometric formula

#height = distance * tan(radians)

#ARGUMENTS
#degrees: The angle of elevation of tree
#distance: The distance from the base of tree (e.g. meters)

#OUTPUT
#The heights of the tree, same units as "distance"

#function for calulating tree height
TreeHeight <- function(degrees, distance){
  radians <- degrees * pi/180
  height <- distance * tan(radians)
  
  return(height)
  
}

args = commandArgs(trailingOnly = TRUE)
print(args)
if (length(args) == 0) {
  #use default if no arguments are provided
  trees = read.csv("../Data/trees.csv", header = TRUE)
  outbase = "default"
} else {
  #provided csv
  trees = read.csv(args[1])
  outbase = args[1]
}

#get base filename strip path and ext
outfile = tools::file_path_sans_ext(basename(outbase))

#contrust output path (suppressing seperators)
outpath = paste0("../Results/", outfile, "_treeheights.csv")

#use dplyr
require(dplyr)
testdf <- trees %>% dplyr::group_by(Species) %>% 
  dplyr::mutate(Tree.Height.m = TreeHeight(Angle.degrees, Distance.m))


#write row names
write.csv(testdf,outpath)

File: Week3/Code/get_TreeHeight.py
#!/usr/bin/env python3

"""

A python script to calculate tree height.

Takes input from an external file and saves tree heights to a new
file.

Output: filename_TreeHeight.txt

"""

__author__ = "Hannah O'Sullivan (h.osullivan18@imperial.ac.uk)"
__appname__ = "Get Tree Heights"
__version__ = "0.0.1"
__license__ = "I do not have one"

#should run using:
#$python get_TreeHeight.py Trees.csv/arg1

#function to caluclate tree heights
#uses distance from base of tree and angle to the top
#trigonometric formula
#height = distance * tan(radians)

#ARGUMENTS
#degrees: The angle of elevation of tree
#distance: The distance from the base of tree (e.g. meters)

#OUTPUT
#The heights of the tree, same units as "distance"

#import useful modules
import sys
import numpy as np
import csv

def TreeHeight(degrees, distance):
    """Calculates tree heights
        Arguments:
        1. degrees: The angle of elevation of the tree
        2. distance: The distance from the base of the tree in meters
        Outputs:
        Tree heights, same units as distance"""
    radians = degrees * np.pi/180
    height = distance * np.tan(radians)
    return(height)


#set default arguments using trees.csvfiles
if len(sys.argv) == 1:
    Trees = "../Data/trees.csv"
    print("No arguments provided, using default.")
elif len(sys.argv) == 2:
    Trees = sys.argv[1]
    print("Calculating tree heights")

#open file
f = open(Trees, 'r')
csvread = csv.reader(f)
csvdata = [i for i in csvread]
c = 0
for i in csvdata:
    c += 1
    if c == 1:
        i.append("Height.m")
        continue
    species = i[0]
    distance = float(i[1])
    degrees = float(i[2])
    height = TreeHeight(distance, degrees)
    i.append(height)
print(csvdata)

outpath ="../Results/{}_treeheights.csv".format(Trees.split("/")[-1].replace(".csv", ""))
with open(outpath,'w') as resultFile:
    wr = csv.writer(resultFile, dialect='excel')
    wr.writerows(csvdata)


#import pandas as pd
# read csv file in
#df = pd.read_csv(Trees)
# make a new column for tree height and vectorize the function TreeHeight
#df["Height.m"] = TreeHeight(degrees=df["Angle.degrees"], distance=df["Distance.m"])
# set path you want to save to
#outpath ="../Results/{}_treeheights.csv".format(Trees.split("/")[-1].replace(".csv", ""))
# write file to csv
#df.to_csv(outpath, index=False)
